### PR TITLE
[4.0] Add a parameter for "back-to-top" button in Cassiopeia

### PR DESCRIPTION
**Pull request by drmenzelit**
Pull Request for Issue #25239 .

### Summary of Changes
Added a parameter to show / hide the back-to-top button in Cassiopeia

### Testing Instructions

Install PR, run npm ci or compile the scss files from Cassiopeia
Go to template, under "advanced" a new switcher should appear
Screenshot_2020-08-21 Templates Edit Style - Joomla Cassiopeia - Administration

### Actual result BEFORE applying this Pull Request

Currently the footer appears when a module has been published on position "footer". And the back-to-top only appears if the footer is visible. Since the footer is part of the grid, it is not easy to move the back-to-top outside the grid.

### Expected result AFTER applying this Pull Request

Now it is possible to decide to have a back-to-top or not. If yes, the footer will be rendered showing the button, independently if there is a module in the position footer.
It is also possible to show only modules (switch back-to-top off) or modules and back-to-top (switch on).

Example with module on footer and back-to-top button

Screenshot_2020-08-21 Home

### CODE REVIEW

**Comment by richard67** [comment]
@drmenzelit Does it fix issue #25239 ?

Update: Yes, it does. Clarified on Glip.

File: language/en-GB/tpl_cassiopeia.ini [hunk 1]
```diff
@@ -9,5 +9,6 @@ TPL_CASSIOPEIA_FLUID_LABEL="Fluid Layout"
 TPL_CASSIOPEIA_LOGO_LABEL="Logo"
 TPL_CASSIOPEIA_STATIC="Static"
 TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
+TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top button"
```
Comment (brianteeman) [inline language/en-GB/tpl_cassiopeia.ini]:
```suggestion
TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top Button"
```

To follow the style guidelines

**Comment by brianteeman** [comment]
Should the sql be updated to include the default setting for this button?

**Comment by brianteeman** [comment]
Is it possible to add scroll-behaviour: smooth - it makes it much nicer than a big jump

See https://css-tricks.com/almanac/properties/s/scroll-behavior/

**Comment by drmenzelit** [comment]
Yes, I was thinking about smooth scrolling, but added in this way:

https://moderncss.dev/pure-css-smooth-scroll-back-to-top/#add-smooth-scrolling

**Comment by brianteeman** [comment]
Its the same isnt it?

File: language/en-GB/tpl_cassiopeia.ini (hunk 1 above)
Comment (drmenzelit) [reply language/en-GB/tpl_cassiopeia.ini]: Changed on new commit

**Comment by drmenzelit** [comment]
with "prefers-reduced-motion: no-preference" we will have the smooth scrolling only if the user has no reduced-motion preferences set

**Comment by brianteeman** [comment]
missed that bit. good spot

**Comment by drmenzelit** [comment]
> Should the sql be updated to include the default setting for this button?

That is not necessary because the default value is set to 0 directly on the field

File: language/en-GB/tpl_cassiopeia.ini (hunk 1 above)
Comment (drmenzelit) [reply language/en-GB/tpl_cassiopeia.ini]: Changed with new commit

**Comment by richard67** [comment]
@drmenzelit I've applied the PR on current 4.0-dev and run `npm ci`and even cleared browser cache and so on, but here it looks strange.

I have a custom html module with some text on footer position, and the back to top button gets the same height as the module has.

backtotop-1

The button works well, and I could live with that, but maybe you can fix it?

**Comment by richard67** [comment]
> > Should the sql be updated to include the default setting for this button?
> 
> That is not necessary because the default value is set to 0 directly on the field

That is ok because the default value is 0. If it was e.g. 1, one would have to save the template style settings for the first time before calling `$this->params->get('backTop')` would return 1.

So maybe we should add it at least to the installation SQL for new installations?

**Comment by drmenzelit** [comment]
> I have a custom html module with some text on footer position, and the back to top button gets the same height as the module has.
> 
> The button works well, and I could live with that, but maybe you can fix it?

I will look at this :-)

**Comment by drmenzelit** [comment]
@richard67 can you please look again? I added height and position to the button.

**Comment by richard67** [comment]
@drmenzelit SCSS code style issue, see report from drone here: https://ci.joomla.org/joomla/joomla-cms/34889/1/24 . It's about the order of properties:

> templates/cassiopeia/scss/blocks/_footer.scss
> 10:5    Expected "height" to come before "border-radius"   order/properties-order
> 11:5    Expected "align-self" to come before "height"      order/properties-order

The "xx:yy" at the beginning of each message is "line number":"position in line".

**Comment by drmenzelit** [comment]
No it should be ok, I found out how to use stylelint ;-)

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 22b39bbe6e8ca1ad24648b8c937742140f677038<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by hans2103** [comment]
I see the following

```
@media screen and (prefers-reduced-motion: no-preference) {
  html,
  body {
    scroll-behavior: smooth;
  }
}
```

Now there is smooth scrolling when the browser setting `prefers-reduced-motion` === `no-preference`
Besides no-preference there is the option `reduced`. https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion
So... when `reduced` is set the `scroll-behavior` will not be set.

What if a new option is introduced?
Wouldn't it be better to turn things around?
```
:root {
  scroll-behavior: smooth;

  // No smooth scroll for people who prefer not to
  @media screen and (prefers-reduced-motion: reduce) {
    scroll-behavior: auto;
  }
}
```

File: language/en-GB/tpl_cassiopeia.ini [hunk 2]
```diff
@@ -9,5 +9,6 @@ TPL_CASSIOPEIA_FLUID_LABEL="Fluid Layout"
 TPL_CASSIOPEIA_LOGO_LABEL="Logo"
 TPL_CASSIOPEIA_STATIC="Static"
 TPL_CASSIOPEIA_STICKY_LABEL="Sticky Header"
+TPL_CASSIOPEIA_BACKTOP_LABEL="Back-to-top Button"
```
Comment (Quy) [inline language/en-GB/tpl_cassiopeia.ini]: Please sort key in alpha order.

Comment (drmenzelit) [reply language/en-GB/tpl_cassiopeia.ini]: Corrected on last commit

**Comment by hans2103** [comment]
I have tested this item :white_check_mark: successfully on ed6dd47c971946c11af84694b5a76e26f15c4ae7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by N6REJ** [comment]
@drmenzelit I would recommend you remove any right padding so it lines up with the container and put it in its own space outside of the footer if.

**Comment by hans2103** [comment]
> @drmenzelit I would recommend you remove any right padding so it lines up with the container and put it in its own space outside of the footer if.

Styling will be taken care of in another PR.

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on ed6dd47c971946c11af84694b5a76e26f15c4ae7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by brianteeman** [comment]
This type of element really should be set o not be printed

`@media print { .back-top{ display: none; } }`

File: templates/cassiopeia/scss/blocks/_global.scss [hunk 3]
```diff
@@ -1,4 +1,12 @@
 // Global
+:root {
+  scroll-behavior: smooth;
+
+  // No smooth scroll for people who prefer not to
```
Comment (brianteeman) [inline templates/cassiopeia/scss/blocks/_global.scss]: Suggesion

```suggestion
  // Disable smooth scroll when the user has set "prefers-reduced-motion"
```

File: language/en-GB/tpl_cassiopeia.ini [hunk 4]
```diff
@@ -4,6 +4,7 @@
 ; Note : All ini files need to be saved as UTF-8
 
 TPL_CASSIOPEIA_BACKTOTOP="Back to Top"
+TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Button"
```
Comment (brianteeman) [inline language/en-GB/tpl_cassiopeia.ini]: technically its a link not a button


```suggestion
TPL_CASSIOPEIA_BACKTOTOP_LABEL="Back-to-top Link"
```

File: templates/cassiopeia/index.php [hunk 5]
```diff
@@ -173,16 +173,15 @@
 	</div>
 	<?php endif; ?>
 
-	<?php if ($this->countModules('footer')) : ?>
+	<?php if ($this->countModules('footer') || ($this->params->get('backTop') == 1)) : ?>
 	<footer class="grid-child container-footer footer">
-		<hr>
-		<p class="float-right">
+		<jdoc:include type="modules" name="footer" style="none" />
```
Comment (infograf768) [inline templates/cassiopeia/index.php:178]: Please test after installing blog sample data.
The menu module is next to the button, so much that they look related. This is really confusing.

**Comment by chmst** [comment]
@infograf768 The template structure with colours and margins / paddings is work in progress.

**Comment by hans2103** [comment]
I have tested this item :white_check_mark: successfully on baac7fb9d4bdf13af9f2b14a718d1b19bbda027b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by hans2103** [comment]
print styling will be taken care of in redesign of Cassiopeia. George mentioned https://twitter.com/dgrammatiko/status/1298231873425092609?s=21 which is interesting <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by alikon** [comment]
i really would like to close this pointing to a real pr , can you  share a link ?

**Comment by hans2103** [comment]
> i really would like to close this pointing to a real pr , can you share a link ?

I hope this is good enough
https://github.com/joomla/cassiopeia/issues

**Comment by chmst** [comment]
I have tested this item :white_check_mark: successfully on baac7fb9d4bdf13af9f2b14a718d1b19bbda027b<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30441">issues.joomla.org/tracker/joomla-cms/30441</a>.</sub>

**Comment by infograf768** [comment]
Tks.

### Reminder
The module displayed in the footer position is real bad and should be corrected.
Screen Shot 2020-08-26 at 10 49 04

**Comment by Scrabble96** [comment]
As I am unable to see the current Cassiopeia go top arrow at all (as of yesterday's Nightly update) - with or without a footer module published - I would like to tentatively suggest this solution for a go to top function that sits outside the footer. Its colours, size and position are controlled by css:

1. In Cassiopeia index.php add this code to below the footer module, above or below the debug:

`<div onClick="topFunction()" id="gotop" class="hide" title="<?php echo Text::_('TPL_CASSIOPEIA_BACKTOTOP'); ?>"><i class="fas fa-arrow-up arrow-up"></i></div>`

2. Add this js to template.js (or user.js for testing purposes):

`/*  Go to Top Link `
`-------------------*/`
`var topbutton = document.getElementById("gotop");`
`window.onscroll = function() {scrollFunction()};`
`function scrollFunction() {`
`  if (document.body.scrollTop > 450 || document.documentElement.scrollTop > 450) {`
`    document.getElementById("gotop").className = "top-link";`
`  } else {`
`    document.getElementById("gotop").className = "top-link hide";`
`  }`
`}`
`function topFunction() {`
`  document.body.scrollTop = 0; // For Safari`
`  document.documentElement.scrollTop = 0; // For Chrome, Firefox, IE and Opera`
`}`

3. Add this css:

`.top-link {`
`  visibility: visible;}`
`.hide {`
` display: none;}`
`.top-link .arrow-up {`
`  align-items: center;`
`  animation-duration: 1s;`
`  background: red;`
`  border: 1px solid white;`
`  bottom: 0;`
`  color: white;`
`  cursor: pointer;`
`  font-size: 20px;`
`  height: 45px;`
`  margin: 0 1em 0 0;`
`  padding-top: .5em;`
`  position: fixed;`
`  left: 0;`
`  text-align: center;`
`  width: 50px;`
`}`
`.top-link .arrow-up:hover {`
`  background: pink;`
`  padding-top: .35em;`
`}`

**Comment by richard67** [comment]
@Scrabble96 You know that you have to switch it on in Cassiopeia's template style settings in the "Advanced" tab?

**Comment by Scrabble96** [comment]
> 
> 
> @Scrabble96 You know that you have to switch it on in Cassiopeia's template style settings in the "Advanced" tab?

I do now! It wasn't there in Protostar, so it's going to have to be made obvious to new users of Cassiopeia that it is switched on or off in the Styles Advanced tab.

**Comment by richard67** [comment]
@Scrabble96 The description of this PR here explains that clearly. I assumed when you are commenting this PR then yopu have read its description, too. And sure, later it has to be described somewhere in documentations.

**Comment by Scrabble96** [comment]
> 
> 
> @Scrabble96 The description of this PR here explains that clearly. I assumed when you are commenting this PR then yopu have read its description, too. And sure, later it has to be described somewhere in documentations.

Sorry. Long day. I'll delete my comment.

**Comment by richard67** [comment]
Sorry if I sounded too harsh. Was a long day for me, too.

**Pull request merged**